sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala

```diff
@@ -276,6 +279,138 @@ object UnresolvedAttribute {
   }
 }
 
+/**
+ * Holds an identifier clause for an attribute that has yet to be resolved.
+ */
+case class UnresolvedAttributeIdentifierClause(expr: Expression)
+  extends Expression with Unevaluable {
+
+  override def children: Seq[Expression] = Seq(expr)
+
+  override def dataType: DataType = throw new UnresolvedException("dataType")
+  override def nullable: Boolean = throw new UnresolvedException("nullable")
+  override lazy val resolved = false
+  final override val nodePatterns: Seq[TreePattern] = Seq(UNRESOLVED_IDENTIFIER_CLAUSE)
+
+  override def prettyName: String = "IDENTIFIER"
+  override def toString: String = {
+    s"'(${children.mkString(", ")})"
+  }
+
+  override protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]):
+  UnresolvedAttributeIdentifierClause = {
+      copy(expr = newChildren.head)
+  }
+}
+
+case class UnresolvedFunctionIdentifierClause(
+                               identExpr: Expression,
+                               arguments: Seq[Expression],
+                               isDistinct: Boolean,
+                               filter: Option[Expression] = None,
+                               ignoreNulls: Boolean = false)
+  extends Expression with Unevaluable {
+
+  override def children: Seq[Expression] = arguments ++ filter.toSeq
+
+  override def dataType: DataType = throw new UnresolvedException("dataType")
+  override def nullable: Boolean = throw new UnresolvedException("nullable")
+  override lazy val resolved = false
+  final override val nodePatterns: Seq[TreePattern] = Seq(UNRESOLVED_IDENTIFIER_CLAUSE)
+
+  override def prettyName: String = identExpr.toString
+  override def toString: String = {
+    val distinct = if (isDistinct) "distinct " else ""
+    s"'${identExpr.toString}($distinct${children.mkString(", ")})"
+  }
+
+  override protected def withNewChildrenInternal
+  (newChildren: IndexedSeq[Expression]): UnresolvedFunctionIdentifierClause = {
+    if (filter.isDefined) {
+      copy(arguments = newChildren.dropRight(1), filter = Some(newChildren.last))
+    } else {
+      copy(arguments = newChildren)
+    }
+  }
+}
+
+object UnresolvedFunctionIdentifierClause {
```

Comment:
optional: it might be simpler to just add this as another constructor inside the above case class instead of creating a separate helper object for this purpose?

Reply:
I think this i unused actually - removed